SwrveSDK/SwrveClient.brs

```diff
@@ -1,1112 +1,1022 @@
+	SwrveSaveStringToPersistence(SwrveConstants().SWRVE_LAST_SESSION_DATE_KEY, date.ToISOString())
+	if NOT dateString = ""
+		SWLogVerbose("Updating last session time:", dateString)
+	end if
+	return dateString
 end function
 
 'Returns the api key'
 function GetAPIKey(swrveClient) as String
-    return m.swrve_config.apikey
+	return m.swrve_config.apikey
 end function
 
 'returns the install date'
 function GetInstallDate(swrveClient) as Integer
-    date = checkOrWriteInstallDate()
-    return date.AsSeconds()
+	date = checkOrWriteInstallDate()
+	return date.AsSeconds()
 end function
 
 'returns the joined date'
 function GetJoinedDate(swrveClient) as Integer
-    date = checkOrWriteJoinedDate()
-    return date.AsSeconds()
+	date = checkOrWriteJoinedDate()
+	return date.AsSeconds()
 end function
 
 function RemoveIAMInClient()
-    m.top.GetScene().dialog = Invalid
+	m.top.GetScene().dialog = Invalid
 end function
 
 
 function ShowIAMInClient()
-    if m.global.swrveShowIAM
-        SWLog("SwrveClient() Swrve has been asked to show an IAM")
-        message = m.global.swrveCurrentIAM
+	if getSwrveNode().showIAM
+		SWLogInfo("SwrveClient() Swrve has been asked to show an IAM")
+		message = getSwrveNode().currentIAM
 
-        d = CreateObject("RoSGNode", "SwrveDialog")
-        d.title = ""
-        d.maxHeight = 200
-        d.graphicWidth = 1385
-        d.graphicHeight = 862
+		d = CreateObject("RoSGNode", "SwrveDialog")
+		d.title = ""
+		d.maxHeight = 200
+		d.graphicWidth = 1385
+		d.graphicHeight = 862
 
-        d.asset_location = SwrveConstants().SWRVE_ASSETS_LOCATION
-        d.iam = message
-        m.top.GetScene().dialog = d
-        d.setFocus = true
+		d.asset_location = SwrveConstants().SWRVE_ASSETS_LOCATION
+		d.iam = message
+		m.top.GetScene().dialog = d
+		d.setFocus = true
 
-        swreveSendImpressionAfterIAMTimer = CreateObject("RoSGNode", "Timer")
-        swreveSendImpressionAfterIAMTimer.duration = 2
-        swreveSendImpressionAfterIAMTimer.ObserveField("fire","SwrveOnDelaySendImpressionEvent")
-        swreveSendImpressionAfterIAMTimer.Control = "start"
-        m.swreveSendImpressionAfterIAMTimer = swreveSendImpressionAfterIAMTimer
-    end if
+		swrveSendImpressionAfterIAMTimer = CreateObject("RoSGNode", "Timer")
+		swrveSendImpressionAfterIAMTimer.duration = 2
+		swrveSendImpressionAfterIAMTimer.ObserveField("fire", "SwrveOnDelaySendImpressionEvent")
+		swrveSendImpressionAfterIAMTimer.Control = "start"
+		m.swrveSendImpressionAfterIAMTimer = swrveSendImpressionAfterIAMTimer
+	end if
 end function
 
 function SwrveOnDelaySendImpressionEvent()
-    if m.swreveSendImpressionAfterIAMTimer <> Invalid
-        m.swreveSendImpressionAfterIAMTimer.control = "stop"
-        m.swreveSendImpressionAfterIAMTimer = Invalid
-    end if
-
-    if(m.global.swrveCurrentIAM <> Invalid)
-        SwrveImpressionEvent(m.global.swrveCurrentIAM)
-    end if
-end function
-
-function SwrveSetupGlobals() as Void
```

Comment:
Since the `SwrveNode` has been updated to contain all these values there is no longer a need to pollute the global scope.

Reply:
Just need to be mindful that this is a major version bump. We advise to call SwrveSetupGlobals() from main in our setup guide.  This may affect child nodes who might be setting Swrve callback functions in their init methods and Swrve node not yet been started/inited.  Would likely need to observe for when Swrve has started.  Are the props now name spaced under a Swrve object in global? Are their advantages to that beside data encapsulation?

Reply:
Yes, I figured this this would result in a bump to then major version. There is follow up to this that I sadly didn't get to but in addition to encapsulation of the data it also makes clean up easier. As now simply deleting that instance on global will also clean up all the fields and result in garbage collection. 

Also if I had more time most actions could have been don't with call functions and such on the `SwrveNode` it self turning the `SwrveNode` into a singleton. This would allow for more data to be cached in memory rather than reading and writing to the registry as much. 

Overall the use of the registry is much higher than I would normally be comfortable with considering there is a limit of 16Kb of space shared with the channel it self.

Reply:
We can take a look at moving to call functions as an improvement in the future.

Re the overall usage of the registry for persistence caching, there are several fields that require persistent storage for the SDK to operate correctly. These would be minimal and take up a small amounts of bytes.

However Campaign and Resource json files could be several kb uncompressed.  These are cached to allow fast load of in app message + save on data transfer costs. The call to get campaigns has an etag and will return empty if nothing is changed, so the SDK will expect to load campaigns and resources from the cache.

Do you know off hand or have experience of what happens if the limit is reached or gone over? Do writes actually stop or is it somewhat a soft limit where things just start to slow down?

Reply:
It's a hard limit. Once the registry is full writes will start to fail. The registry is definitely not designed for caching. 

One thing that you could look at leveraging is cachefs. This is a ￼ volatile file system stored in ram. It can live between channel runs, though not ￼guaranteed, and is dumped as needed by the os in order to run other channels and when the device restarts. 

Caching on Roku is ￼extremely difficult and limited. Most caching is done on a per run bases for this reason.

Reply:
To add to this conversation a system should be created that adds some form of TTL or basic cleanup to the registry. This is a bit more complicated topic but the issue is the registry usage in its current form has a leak. 

For example:
A channel that identifies users by profile rather then by the account the profile belongs to can results in many sections of stale data being created if a user where to create and delete many profiles. 

The most common way I have see. This handed is with Least Recently Used (LRU) cache style.

Reply:
Thanks @chrisdp we will look into the cachefs for larger items and providing an API to clear cached items based on user